sunpy/data/data_manager/storage.py

```diff
@@ -0,0 +1,176 @@
+"""
+Storage module contains the abstract implementation of storage
+for `sunpy.data.data_manager.Cache` and a concrete implementation
+using sqlite.
+"""
+import os
+import sqlite3
+from abc import ABCMeta, abstractmethod
+from contextlib import contextmanager
+from pathlib import Path
+
+
+class StorageProviderBase(metaclass=ABCMeta):
+    """
+    Base class for remote data manager storage providers.
+    """
+    @abstractmethod
+    def find_by_key(self, key, value):
+        """
+        Returns the file details if value coresponding to the key
+        found in storage. Returns `None` if hash not found.
+
+        Parameters
+        ----------
+        key: `str`
+            The key/coloumn name of the field.
+        value: `str`
+            The value associated with the key of the entry.
+
+        Returns
+        -------
+        `dict` or `None`
+            `dict` contains the details of the file. `None` if hash not found.
```

Comment:
Should we specify a schema for this dict somehow? It's a really critical part of this interface which this ABC can't specify.

(There is something new in the typing module we could use in 3.8 but we can't use that yet, not sure what the best way to do it at the moment would be? (Also we probably want runtime checking, maybe a custom class which implements `Mapping`?))

Reply:
I remember looking into dict schemas in typing when I was writing this. Apparently, typing is not powerful enough to do it ATM.

I found this though https://pypi.org/project/jsonschema/, but adding a dependency just for this did not make sense.

Reply:
Yeah I think the only practical solution would be to implement a specific class for it. Not really sure it's worth it.